feat: implement contract migration to add projects to pagination

masch opened the pull request:
# 🚀 Migration of Existing Projects to Pagination System

## Description
This PR implements the `MigrationTrait` to allow administrators to manually link existing orphaned projects (projects that exist in persistent storage but are not yet tracked in the pagination lists) into the contract's pagination system.

## 🔗 Related Issue
Closes #331

## 🛠️ Key Changes

### Core Logic
- **Name-based Migration**: The implementation in allows passing plain project names (`Vec<String>`).
- **Internal Hashing**: Automatically calculates project keys from names using Keccak256 to match the registration system.
- **Improved Duplication Check**: Replaced manual loops with idiomatic `.contains()` checks for project keys.

### Refactoring & Security
- **Shared Constants**: Moved `MAX_PROJECTS_PER_PAGE` to `types.rs` to unify pagination logic across `contract_versioning` and `contract_migration`.
- **Administrative Controls**: 
  - Integrated `auth_admin` check to restrict usage to authorized administrators.
  - Added `require_not_paused` check to prevent migrations during maintenance modes.

## 🧪 Verification Results

### Automated Tests
A new test suite [`test_migration.rs`](contracts/tansu/src/tests/test_migration.rs) was added to verify all edge cases:
- test_add_projects_to_pagination_previously_registered_project: Verifies migration of orphaned keys and duplicate prevention.
- test_add_projects_to_pagination_unregistered_project: Ensures non-existent project names are ignored.
- test_add_projects_to_pagination_paused_error: Validates respect for the contract's paused state.
- test_add_projects_to_pagination_unauthorized_admin_error: Confirms only admins can execute the migration.

tupui commented on contracts/tansu/src/contract_versioning.rs:
```diff
@@ -50,15 +47,7 @@ impl VersioningTrait for Tansu {
-        name.copy_into_slice(&mut slice[..str_len]);
-        let name_b = Bytes::from_slice(&env, &slice[0..str_len]);
-
+        let name_b = crate::get_name_bytes(&env, &name);
```
No don't do any refactoring for a migration. Migrations are one-off scripts and must not drive the business logic. Same goes for tests. If we need to re-do something, it's perfectly fine. The other issue is that doing so breaks code lineage.

masch replied:
Reverted, I was trying to avoid duplication.

tupui reviewed (changes requested):
Thanks for working on this. We can drastically simplify the approach. I left some comments.

A quick note on AI. Avoid letting it write comments for PR descriptions, or even in general. It never works besides being overly confident that it did things perfectly and adding useless emoji all over the place 😅

masch commented on contracts/tansu/src/lib.rs:
```diff
@@ -208,3 +213,40 @@ fn validate_contract(env: &Env, contract: &types::Contract) {
+///
+/// # Panics
+/// * If the name is longer than 15 bytes
+pub fn get_project_key(env: &Env, name: &String) -> Bytes {
```
I did an extra refactor earlier to avoid duplicating the function that generates project keys. I've reverted the commit.

masch commented:
> Thanks for working on this. We can drastically simplify the approach. I left some comments.
> 
> A quick note on AI. Avoid letting it write comments for PR descriptions, or even in general. It never works besides being overly confident that it did things perfectly and adding useless emoji all over the place 😅

I manually added the comments in the test to explain the logic, and I followed the existing pattern for the trait function definitions. Regarding the PR comments, you’re right—those were AI-generated suggestions.

Let me know if this looks like the right approach for the project! 🎅

masch commented:
I’ve added support for multiple pages in the migration. I also added a check to see if the project name already exists on any page; although that’s a very rare edge case, it’s a good safety measure to have.
